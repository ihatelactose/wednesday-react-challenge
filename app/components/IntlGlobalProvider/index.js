// eslint-disable-next-line
import React from 'react';
import { useIntl } from 'react-intl';

// 'intl' service singleton reference
let intl;

export function IntlGlobalProvider({ children }) {
  intl = useIntl(); // Keep the 'intl' service reference
  return children;
}

// setter function to set intl value inside tests
export const setIntl = (intlValue) => {
  intl = intlValue;
};

export const translate = (id, values = {}) => intl.formatMessage({ id }, values);
