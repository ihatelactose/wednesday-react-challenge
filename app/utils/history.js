import { createBrowserHistory } from 'history';
import { isProd, isUAT } from './index';

export function getBaseName() {
  if (isProd()) {
    // GH Pages
    return '/react-template';
  }
  if (isUAT()) {
    return `/${process.env.BRANCH_NAME}`;
  }

  return '';
}

const history = createBrowserHistory({ basename: getBaseName() });
export default history;
