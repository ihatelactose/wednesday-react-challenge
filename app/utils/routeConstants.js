export default {
  repos: {
    route: '/',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  },
  artists: {
    route: '/artists',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  }
};
