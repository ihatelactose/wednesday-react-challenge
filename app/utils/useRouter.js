import { useMemo } from 'react';
import { useParams, useLocation, useHistory, useRouteMatch } from 'react-router-dom';
import queryString from 'query-string';

export function useRouter() {
  const params = useParams();
  const location = useLocation();
  const history = useHistory();
  const match = useRouteMatch();

  // Memoize so as to return a new object if and only if something changes
  return useMemo(() => {
    return {
      push: history.push,
      replace: history.replace,
      pathname: location.pathname,
      query: {
        ...queryString.parse(location.search),
        ...params
      },
      match,
      location,
      history
    };
  }, [params, location, history, match]);
}
