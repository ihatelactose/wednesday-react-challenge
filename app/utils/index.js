import find from 'lodash/find';
import get from 'lodash/get';
import routeConstants from './routeConstants';

/**
 * Get details of the current route from the route config.
 * If location.pathname doesn't match any entry in the routeConstants it will return null
 * @author mac
 * @date 2020-05-07
 * @param {any} location
 * @returns {any}
 */
export const getCurrentRouteDetails = (location) => {
  if (!get(location, 'pathname')) {
    return null;
  }
  const route = find(
    Object.keys(routeConstants),
    (key) => routeConstants[key].route === location.pathname || `${routeConstants[key].route}/` === location.pathname
  );
  if (route) {
    return routeConstants[route];
  }
  return null;
};
export const mapKeysDeep = (obj, fn) =>
  Array.isArray(obj)
    ? obj.map((val) => mapKeysDeep(val, fn))
    : typeof obj === 'object'
    ? Object.keys(obj).reduce((acc, current) => {
        const key = fn(current);
        const val = obj[current];
        acc[key] = val !== null && typeof val === 'object' ? mapKeysDeep(val, fn) : val;
        return acc;
      }, {})
    : obj;

export const isLocal = () => {
  try {
    const local = JSON.parse(process.env.IS_LOCAL);
    return typeof local === 'boolean' && local;
  } catch {
    // continue regardless of error
  }
  return false;
};

export function isUAT() {
  return process.env.ENVIRONMENT_NAME === 'development' && process.env.NODE_ENV === 'production';
}

export function isProd() {
  return process.env.ENVIRONMENT_NAME === 'production' && process.env.NODE_ENV === 'production';
}
