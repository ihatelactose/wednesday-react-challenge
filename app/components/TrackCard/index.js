/**
 *
 * Track Card
 *
 */
import { Card, Image, Button } from 'antd';
import { isEmpty } from 'lodash';
import PropTypes from 'prop-types';
import React, { useEffect, useRef, useState } from 'react';
import styled from 'styled-components';
import If from '@components/If/index';
import { T } from '@components/T/index';
import { useRouter } from '@app/hooks/useRouter';

const CustomCard = styled(Card)`
  && {
    margin: 20px 0;
  }
`;

const Flex = styled.div`
  display: flex;
  align-items: center;
  margin-bottom: 10px;
`;

const CustomButton = styled(Button)`
  margin-bottom: ${(props) => props.mb}px;
  margin-top: ${(props) => props.mt}px;
  margin-left: ${(props) => props.ml}px;
  margin-right: ${(props) => props.mr}px;
`;

export function TrackCard({
  artistName,
  artistViewUrl,
  artworkUrl100,
  collectionName,
  collectionPrice,
  currency,
  releaseDate,
  previewUrl,
  currentlyPlaying,
  trackId,
  dispatchCurrentlyPlaying
}) {
  const [isPlaying, setIsPlaying] = useState(false);
  const router = useRouter();
  const audioRef = useRef(null);

  useEffect(() => {
    if (audioRef) {
      if (currentlyPlaying !== trackId) {
        audioRef.current.pause();
        setIsPlaying(false);
      }
    }
  }, [currentlyPlaying]);

  function handleDetailsClick(evt) {
    evt.preventDefault();
    router.push(`/track/${trackId}`);
  }

  function handleOnPlay(evt) {
    evt.preventDefault();
    if (isPlaying) {
      audioRef.current.pause();
      setIsPlaying(false);
      dispatchCurrentlyPlaying(null);
    } else {
      audioRef.current.play();
      setIsPlaying(true);
      dispatchCurrentlyPlaying(trackId);
    }
  }

  return (
    <CustomCard data-testid="artist-card">
      <Flex>
        <If
          condition={!isEmpty(artworkUrl100)}
          otherwise={<T data-testid="artwork-url-unavailable" id="artwork_url_unavailable" />}
        >
          <div style={{ marginRight: '10px' }}>
            <Image src={artworkUrl100} alt={collectionName} width={100} />
          </div>
        </If>
        <If
          condition={!isEmpty(collectionName)}
          otherwise={<T data-testid="collection-name-unavailable" id="collection_name_unavailable" />}
        >
          <T data-testid="collection-name" id="collection_name" values={{ collectionName }} />
        </If>
      </Flex>
      <If
        condition={!isEmpty(previewUrl)}
        otherwise={<T data-testid="preview-audio-unavailable" id="preview_audio_unavailable" />}
      >
        <Button data-testid="preview-audio-button" onClick={handleOnPlay} type="primary">
          {isPlaying ? 'Pause preview' : 'Play Preview'}
        </Button>
        <audio data-testid="preview-audio" ref={audioRef} src={previewUrl} />
      </If>
      <If
        condition={!isEmpty(artistName)}
        otherwise={<T data-testid="artist-name-unavailable" id="artist_name_unavailable" />}
      >
        <T data-testid="artist-name" id="artist_name" values={{ artistName }} />
      </If>
      <If condition={!isEmpty(artistViewUrl)}>
        <a href={artistViewUrl} rel="noreferrer" target="_blank">
          Check out the artist!
        </a>
      </If>
      <If
        condition={!isNaN(collectionPrice)}
        otherwise={<T data-testid="collection-price-unavailable" id="collection_price_unavailable" />}
      >
        <T data-testid="collection-price" id="collection_price" values={{ currency, collectionPrice }} />
      </If>
      <If
        condition={!isEmpty(releaseDate)}
        otherwise={<T data-testid="release-date-unavailable" id="release_date_unavailable" />}
      >
        <T
          data-testid="release-date"
          id="release_date"
          values={{ releaseDate: new Date(releaseDate).toDateString() }}
        />
      </If>
      <CustomButton onClick={handleDetailsClick} mt={20}>
        <T id="track_card_default_details" />
      </CustomButton>
    </CustomCard>
  );
}

TrackCard.propTypes = {
  artistName: PropTypes.string,
  artistViewUrl: PropTypes.string,
  artworkUrl100: PropTypes.string,
  collectionName: PropTypes.string,
  collectionPrice: PropTypes.number,
  country: PropTypes.string,
  currency: PropTypes.string,
  releaseDate: PropTypes.string,
  previewUrl: PropTypes.string,
  currentlyPlaying: PropTypes.number,
  dispatchCurrentlyPlaying: PropTypes.func,
  trackId: PropTypes.number
};

export default TrackCard;
