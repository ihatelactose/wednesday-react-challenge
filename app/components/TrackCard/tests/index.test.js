/**
 *
 * Tests for TrackCard
 *
 */

import React from 'react';
import TrackCard from '../index';
import { renderWithIntl, timeout } from '@app/utils/testUtils';
import { fireEvent } from '@testing-library/dom';
import { translate } from '@app/components/IntlGlobalProvider/index';
import { BrowserRouter } from 'react-router-dom';

describe('<TrackCard />', () => {
  jest.spyOn(window.HTMLMediaElement.prototype, 'pause').mockImplementation(() => {});
  jest.spyOn(window.HTMLMediaElement.prototype, 'play').mockImplementation(() => {});

  it('should render and match with the snapshot', () => {
    const baseComponent = renderWithIntl(
      <BrowserRouter>
        <TrackCard />
      </BrowserRouter>
    );
    expect(baseComponent).toMatchSnapshot();
  });

  it('contains a TrackCard component', () => {
    const { getAllByTestId } = renderWithIntl(
      <BrowserRouter>
        <TrackCard />
      </BrowserRouter>
    );
    expect(getAllByTestId('artist-card').length).toBe(1);
  });

  it('renders the artists details inside the mounted component', () => {
    const artistName = 'Stromae';
    const artworkUrl100 =
      'https://is1-ssl.mzstatic.com/image/thumb/Music116/v4/71/3a/d3/713ad3a8-54b1-0c6e-f978-71fc540daed5/source/100x100bb.jpg';
    const artistViewUrl = 'https://music.apple.com/us/artist/stromae/270954369?uo=4';
    const collectionName = 'Multitude';
    const collectionPrice = 14.99;
    const currency = 'USD';
    const releaseDate = '2022-01-09T12:00:00Z';

    const { getByTestId } = renderWithIntl(
      <BrowserRouter>
        <TrackCard
          artistName={artistName}
          artistViewUrl={artistViewUrl}
          artworkUrl100={artworkUrl100}
          collectionName={collectionName}
          collectionPrice={collectionPrice}
          currency={currency}
          releaseDate={releaseDate}
        />
      </BrowserRouter>
    );

    expect(getByTestId('collection-name')).toHaveTextContent(collectionName);
    expect(getByTestId('artist-name')).toHaveTextContent(artistName);
    expect(getByTestId('collection-price')).toHaveTextContent(collectionPrice);
    expect(getByTestId('release-date')).toHaveTextContent(new Date(releaseDate).toDateString());
  });

  it('renders the error messages if the artist details are not passed', () => {
    const artistName = translate('artist_name_unavailable');
    const artworkUrl100 = translate('artwork_url_unavailable');
    const collectionName = translate('collection_name_unavailable');
    const collectionPrice = translate('collection_price_unavailable');
    const releaseDate = translate('release_date_unavailable');

    const { getByTestId } = renderWithIntl(
      <BrowserRouter>
        <TrackCard />
      </BrowserRouter>
    );

    expect(getByTestId('artwork-url-unavailable')).toHaveTextContent(artworkUrl100);
    expect(getByTestId('artist-name-unavailable')).toHaveTextContent(artistName);
    expect(getByTestId('collection-name-unavailable')).toHaveTextContent(collectionName);
    expect(getByTestId('collection-price-unavailable')).toHaveTextContent(collectionPrice);
    expect(getByTestId('release-date-unavailable')).toHaveTextContent(releaseDate);
  });

  it('plays and pauses the song when the play/pause preview button is clicked', () => {
    const clickSpy = jest.fn();
    const previewUrl =
      'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview122/v4/a8/e1/fd/a8e1fd7c-b8af-3c54-6665-26f7093c29b6/mzaf_16989964717297819293.plus.aac.p.m4a';
    const trackId = 12345;
    const { getByTestId } = renderWithIntl(
      <BrowserRouter>
        <TrackCard previewUrl={previewUrl} trackId={trackId} dispatchCurrentlyPlaying={clickSpy} />
      </BrowserRouter>
    );

    const audioPreviewButton = getByTestId('preview-audio-button');

    fireEvent.click(audioPreviewButton);
    expect(clickSpy).toBeCalledWith(trackId);

    fireEvent.click(audioPreviewButton);
    expect(clickSpy).toBeCalledWith(null);
  });

  it('should auto-pause the track if the currentlyPlaying trackId is not equal to the selected trackId', async () => {
    const trackId = 12345;
    const currentlyPlaying = 54321;
    const previewUrl =
      'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview122/v4/a8/e1/fd/a8e1fd7c-b8af-3c54-6665-26f7093c29b6/mzaf_16989964717297819293.plus.aac.p.m4a';
    const { getByTestId } = renderWithIntl(
      <BrowserRouter>
        <TrackCard previewUrl={previewUrl} trackId={trackId} currentlyPlaying={currentlyPlaying} />
      </BrowserRouter>
    );

    await timeout(500);
    const audioElement = getByTestId('preview-audio');

    expect(audioElement.src).toEqual(previewUrl);
  });
});
