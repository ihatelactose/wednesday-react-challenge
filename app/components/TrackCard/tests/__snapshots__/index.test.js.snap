// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackCard /> should render and match with the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ant-card ant-card-bordered TrackCard__CustomCard-sc-10bk196-0 bcUGih"
        data-testid="artist-card"
      >
        <div
          class="ant-card-body"
        >
          <div
            class="TrackCard__Flex-sc-10bk196-1 eJxPar"
          >
            <p
              class="T__StyledText-gjlic1-0 kYcyRq"
              data-testid="artwork-url-unavailable"
            >
              The artwork for this track or artist is not available.
            </p>
            <p
              class="T__StyledText-gjlic1-0 kYcyRq"
              data-testid="collection-name-unavailable"
            >
              Sorry this collection is not available.
            </p>
          </div>
          <p
            class="T__StyledText-gjlic1-0 kYcyRq"
            data-testid="preview-audio-unavailable"
          >
            Sorry preview audio for this track is not available.
          </p>
          <p
            class="T__StyledText-gjlic1-0 kYcyRq"
            data-testid="artist-name-unavailable"
          >
            Sorry this artist is not available.
          </p>
          <p
            class="T__StyledText-gjlic1-0 kYcyRq"
            data-testid="collection-price-unavailable"
          >
            Sorry price is not available yet.
          </p>
          <p
            class="T__StyledText-gjlic1-0 kYcyRq"
            data-testid="release-date-unavailable"
          >
            Release date for this collection is not available.
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ant-card ant-card-bordered TrackCard__CustomCard-sc-10bk196-0 bcUGih"
      data-testid="artist-card"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="TrackCard__Flex-sc-10bk196-1 eJxPar"
        >
          <p
            class="T__StyledText-gjlic1-0 kYcyRq"
            data-testid="artwork-url-unavailable"
          >
            The artwork for this track or artist is not available.
          </p>
          <p
            class="T__StyledText-gjlic1-0 kYcyRq"
            data-testid="collection-name-unavailable"
          >
            Sorry this collection is not available.
          </p>
        </div>
        <p
          class="T__StyledText-gjlic1-0 kYcyRq"
          data-testid="preview-audio-unavailable"
        >
          Sorry preview audio for this track is not available.
        </p>
        <p
          class="T__StyledText-gjlic1-0 kYcyRq"
          data-testid="artist-name-unavailable"
        >
          Sorry this artist is not available.
        </p>
        <p
          class="T__StyledText-gjlic1-0 kYcyRq"
          data-testid="collection-price-unavailable"
        >
          Sorry price is not available yet.
        </p>
        <p
          class="T__StyledText-gjlic1-0 kYcyRq"
          data-testid="release-date-unavailable"
        >
          Release date for this collection is not available.
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
